src/main/java/io/confluent/copycat/hdfs/WALFile.java

```diff
@@ -0,0 +1,953 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.confluent.copycat.hdfs;
+
+import org.apache.commons.io.Charsets;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.ChecksumException;
+import org.apache.hadoop.fs.CommonConfigurationKeys;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.Syncable;
+import org.apache.hadoop.io.DataInputBuffer;
+import org.apache.hadoop.io.DataOutputBuffer;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.hadoop.io.VersionMismatchException;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.io.serializer.Deserializer;
+import org.apache.hadoop.io.serializer.SerializationFactory;
+import org.apache.hadoop.io.serializer.Serializer;
+import org.apache.hadoop.util.Options;
+import org.apache.hadoop.util.Time;
+
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.rmi.server.UID;
+import java.security.MessageDigest;
+import java.util.Arrays;
+
+public class WALFile {
+
+  private static final Log log = LogFactory.getLog(WALFile.class);
+  private static final byte INITIAL_VERSION = (byte) 0;
+  private static byte[] VERSION = new byte[]{
+      (byte) 'W', (byte) 'A', (byte) 'L', INITIAL_VERSION
+  };
+
+  private static final int SYNC_ESCAPE = -1;      // "length" of sync entries
+  private static final int SYNC_HASH_SIZE = 16;   // number of bytes in hash
+  private static final int SYNC_SIZE = 4 + SYNC_HASH_SIZE; // escape + hash
+
+  /**
+   * The number of bytes between sync points.
+   */
+  public static final int SYNC_INTERVAL = 100 * SYNC_SIZE;
+
+  private WALFile() {}
+
+  public static Writer createWriter(Configuration conf, Writer.Option... opts) throws IOException {
+    return new Writer(conf, opts);
+  }
+
+
+  public static class Writer implements Closeable, Syncable {
+
+    private Configuration conf;
+    private FSDataOutputStream out;
+    private DataOutputBuffer buffer = new DataOutputBuffer();
+    boolean ownOutputStream = true;
+    private boolean appendMode;
+    protected Serializer keySerializer;
+    protected Serializer valSerializer;
+
+    // Insert a globally unique 16-byte value every few entries, so that one
```

Comment:
What should I be comparing this to? Is this basically SequenceFile but with a few modifications? If it is based off of something else, it'd be easier to review if I could diff it against that.